crates/rustc_utils/src/cache.rs

```diff
@@ -60,6 +85,49 @@ impl<In, Out> Default for CopyCache<In, Out> {
   }
 }
 
+/// This cache alters the [`Self::get`] method signature to return
+/// an [`Option`] of a reference. In particular the method will return [`None`]
+/// if it is called *with the same key* while computing a construction function
+/// for that key.
+pub struct RecursionBreakingCache<In, Out>(RefCell<HashMap<In, Option<Pin<Box<Out>>>>>);
```

Comment:
Rather than duplicating half the logic and type structure between `Cache` and `RecursionBreakingCache`, I would just add a method `get_opt` to `Cache` that has the logic you added, and then change `get` to be `get_opt().unwrap()`.

Reply:
I share your concern about duplication, but I would disagree on the naming part. It is too easy in my opinion to just use `get` when it is available. I would prefer users pick at the type level wether they anticipate recursion and then possibly have e.g. a `get_unsafe` that does the `unwrap`. A `get_opt` seems "too weak" to me.

Reply:
I guess your point os to just make `Cache` recursion resistant by default though? That I would agree with. Maybe the fact that `get` may panic could be communicated more clearly though. Something akin to what `expect` and `unwrap` communicate to the user.